osc/osc.go

```diff
@@ -52,17 +76,22 @@ var _ Packet = (*Bundle)(nil)
 // Client enables you to send OSC packets. It sends OSC messages and bundles to
 // the given IP address and port.
 type Client struct {
-	ip    string
-	port  int
-	laddr *net.UDPAddr
+	ip              string
+	port            int
+	laddr           *net.UDPAddr
+	laddrTCP        *net.TCPAddr
+	networkProtocol NetworkProtocol
 }
 
 // Server represents an OSC server. The server listens on Address and Port for
 // incoming OSC packets and bundles.
 type Server struct {
-	Addr        string
-	Dispatcher  Dispatcher
-	ReadTimeout time.Duration
+	Addr            string
+	Dispatcher      Dispatcher
+	ReadTimeout     time.Duration
+	networkProtocol NetworkProtocol
+	udpConnection   net.PacketConn
+	tcpListener     net.Listener
```

Comment:
It would be great to abstract `udpConnection` and `tcpListerner` out a little so that there is a field on the `Server` that is populated by an interface that is common to both TCP and UDP. Right off the top of my head I can't quite work it out, but I believe we would be able to do it.
I'm going to take a deeper look at this tomorrow to see what if we can work something out.

Reply:
@glynternet I played around with this a bit, but I ran into an issue.

The `NewClient` method takes 2 parameters, `host string, port int`. Go doesn't have method overloading, so I can't define `NewClient(host string, port int, protocol NetworkProtocol)` because there is already a method called `NewClient`. So I figured that to change the network protocol, you would have to create a client first and then use `SetNetworkProtocol` to set the protocol.

That happens to work OK with my current implementation on this branch because all the fields are there (i.e. both `udpConnection` and `tcpListener`), so it doesn't matter what order you call i.e. `SetLocalAddr` vs. `SetNetworkProtocol`.

I experimented with creating a Transport interface with implementations UDPTransport and TCPTransport. The individual Transport implementations then became a place to put the transport-specific properties like `udpConnection` and `tcpListener`. `SetLocalAddr` just called e.g. `client.transport.SetLocalAddr`. The problem with that approach, though, is that because of the setup above, **now it matters which order you call the client and server setter methods**. If you called `SetLocalAddr(..., ...)` followed by `SetNetworkProtocol(TCP)`, the call to `SetNetworkProtocol(TCP)` would blow away the address and port information that you just set, because it was stored on the old network protocol.

Maybe it would be cleaner to add a second constructor, `NewClientTCP`, and remove `SetNetworkProtocol`. There could also be a `NewClientUDP` and `NewClient` could delegate to `NewClientUDP` for backwards compatibility. I'll try that next.

Reply:
Yikes, this is ending up being really difficult. I started down the path of doing an insane amount of refactoring. I think I'm inclined to leave this alone, if that's OK. I'm worried that if I tried to "fix" this, it would end up causing a much larger set of changes than I'd intended.

I'd be happy for anyone else to try their hand at refactoring to use an interface, if anyone is interested.

Reply:
I've pushed my WIP attempt at this to a branch, for the curious. It's not in a working state, but you can at least see what I tried and the explosion of changes it was beginning to require.

https://github.com/daveyarwood/go-osc/compare/tcp-support...wip-refactor-transport-interface